libnavigation-base/src/main/java/com/mapbox/navigation/base/options/NavigationOptions.kt

```diff
@@ -54,6 +51,46 @@ data class NavigationOptions(
      */
     fun toBuilder() = builder
```

Comment:
are we going to fix `toBuilder()` methods in a separate PR?
the next check fails:
```
@Test
fun reuseChangedBuilder() {
    val builder = NavigationOptions.Builder(context)
    val options = builder.build()
    builder.accessToken("pk.123")

    assertEquals(options, options.builder.build())
}
```
as @LukasPaczos  mentioned we should build the `Builder` object from the scratch
`fun toBuilder() = Builder(required).apply { non-required }` 
and remove `val builder: Builder`

Reply:
:+1: would be great to address in this PR

Reply:
the reuse test passes in this PR. the new toBuilder pattern is a matter of preference

Reply:
here's the test https://github.com/mapbox/mapbox-navigation-android/pull/3364/files#diff-cd3a8f1153e5b73755727f75b8dacec6R98-R112

will one of you follow up with a new toBuilder pattern with the justification?

Reply:
on one end, we have to maintain the toBuilder.apply function
on the other, we need to remember not to add the builder to the equals/hash code generate

is there a tie breaker? having a builder within the constructor can [cause memory leaks](https://github.com/mapbox/mapbox-navigation-android/pull/3023) could be a good reason

Reply:
> here's the test https://github.com/mapbox/mapbox-navigation-android/pull/3364/files#diff-cd3a8f1153e5b73755727f75b8dacec6R98-R112

This test doesn't check what we're talking about here. It's not about taking two identical builders and checking if they are the same, it's about modifying a builder. We cannot cache a reference to the builder, because it can result in inconsistencies if the builder has been modified. This is explained here https://github.com/mapbox/mapbox-navigation-android/issues/2709#issuecomment-658863117.

Reply:
Oh wait, sorry I guess I didn't and don't understand. Copied this test in, and sure enough it fails.

``` kotlin
@Test
fun reuseChangedBuilder() {
    val builder = NavigationOptions.Builder(context)
    val options = builder.build()
    builder.accessToken("pk.123")

    assertEquals(options, options.builder.build())
}
```

But I'm actually confused. Why should they be equal? The access token is different, it has only changed on the builder but not the prebuilt options?

Reply:
But now I'm seeing the point you two are making, sorry I'm slow. We don't want to be able to change the builder of other options. But if you want to capture that case, it needs to look like this:

``` kotlin
    @Test
    fun reuseChangedBuilder() {
        val builder = NavigationOptions.Builder(context)
        val options = builder.build()
        builder.accessToken("pk.123")

        assertNotEquals(options.toBuilder().build(), options.toBuilder().build())
    }
```

Reply:
builder shouldn't be cached inside `NavigationOptions ` or any other class.
when we call `options.toBuilder()` we should receive a new, just created, instance.
and `assertEquals(options, options.builder.build())` should always be true

Reply:
This is a hypothetical case that would not be blocking. We're also supporting something we're not planning on advertising. That makes this maintenance overkill in my opinion. 

But I've shared that numerous times. Adding the new toBuilder 🙃

Reply:
I didn't think it's fair to call this case hypothetical. In the end, it would be considered a bug if someone would run into this scenario.

Reply:
The reason I called it hypothetical and not blocking, is because there are workarounds. Use a new builder and you won't run into the issue. This issue only applies to those saving their builders.

But you two are correct, I didn't see the case at first. Added and updated!